### ISSUE TITLE
Slate: high-degree problems surpasses Eigen's stack allocation limit for fixed matrix types

### ISSUE THREAD

**Issue by thomasgibson** [firedrakeproject/firedrake#1221]
I'm logging this issue here for now and will gradually update as solutions are found.

This came up while running some HDG tests in three-dimensions. I've added a code snippet of the Slate-generated code for the static condensation of a three-dimensional HDG problem (degree 5 on tetrahedra).
```cpp
 static inline void compile_slate (double  A2[84][84] , double *  coords , double *  w_0 , int8_t *  arg_cell_facets )
{
  int8_t (*cell_facets)[2] = (int8_t (*)[2])arg_cell_facets;
  /* Declare and initialize */
  Eigen::Matrix<double, 308, 308, Eigen::RowMajor>  T0 ;
  T0.setZero();
  /* Assemble local tensors */
  subkernel0_cell_to_00_cell_integral_otherwise(T0.block<168, 168>(0, 0), coords);
  subkernel0_cell_to_01_cell_integral_otherwise(T0.block<168, 56>(0, 168), coords);
  subkernel0_cell_to_10_cell_integral_otherwise(T0.block<56, 168>(168, 0), coords);
  subkernel0_cell_to_11_cell_integral_otherwise(T0.block<56, 56>(168, 168), coords);
  /* Loop over cell facets */
  
  for (unsigned int  i0  = 0; i0 < 4; i0 += 1)
  {
    if (cell_facets[i0][0] == 0) {
      subkernel0_exterior_facet_to_02_exterior_facet_integral_otherwise(T0.block<168, 84>(0, 224), coords, w_0, &i0);
      subkernel0_exterior_facet_to_10_exterior_facet_integral_otherwise(T0.block<56, 168>(168, 0), coords, w_0, &i0);
      subkernel0_exterior_facet_to_11_exterior_facet_integral_otherwise(T0.block<56, 56>(168, 168), coords, w_0, &i0);
      subkernel0_exterior_facet_to_12_exterior_facet_integral_otherwise(T0.block<56, 84>(168, 224), coords, w_0, &i0);
      
    }
     if (cell_facets[i0][0] == 1) {
      subkernel0_interior_facet_to_02_exterior_facet_integral_otherwise(T0.block<168, 84>(0, 224), coords, w_0, &i0);
      subkernel0_interior_facet_to_10_exterior_facet_integral_otherwise(T0.block<56, 168>(168, 0), coords, w_0, &i0);
      subkernel0_interior_facet_to_11_exterior_facet_integral_otherwise(T0.block<56, 56>(168, 168), coords, w_0, &i0);
      subkernel0_interior_facet_to_12_exterior_facet_integral_otherwise(T0.block<56, 84>(168, 224), coords, w_0, &i0);
      subkernel0_interior_facet_to_20_exterior_facet_integral_otherwise(T0.block<84, 168>(224, 0), coords, w_0, &i0);
      subkernel0_interior_facet_to_21_exterior_facet_integral_otherwise(T0.block<84, 56>(224, 168), coords, w_0, &i0);
      subkernel0_interior_facet_to_22_exterior_facet_integral_otherwise(T0.block<84, 84>(224, 224), coords, w_0, &i0);
      
    }
     
  }
  Eigen::PartialPivLU<Eigen::Matrix<double, 224, 224, Eigen::RowMajor> > dec1((T0).block<224, 224>(0, 0));
  /* Map eigen tensor into C struct */
  Eigen::Map<Eigen::Matrix<double, 84, 84, Eigen::RowMajor> > T2((double *)A2);
  /* Linear algebra expression */
  T2 += ((T0).block<84, 84>(224, 224)) + (-((T0).block<84, 224>(224, 0)) * dec1.inverse() * ((T0).block<224, 84>(0, 224)));
  
}
```
As you can see, the local tensors are quite large. Compiling the full generated code results in the following compilation error: [hdg_sc.log](https://github.com/firedrakeproject/firedrake/files/1937830/hdg_sc.log)

TLDR: The error `‘OBJECT_ALLOCATED_ON_STACK_IS_TOO_BIG’` is raised. The guilty memeber in particular is the 308 x 308 local tensor. This is due to the fact (I believe) that fixed-sized objects are being allocated with too big dimensions. One possible solution I think is to increase stack allocation limit via `EIGEN_STACK_ALLOCATION_LIMIT`. I don't think this is desirable though... Anyone else know a workout? I'll continue searching for possible solutions.

**Comment by wence-** [comment]
OK, so it looks like if you statically declare matrix sizes then they are stack allocated.  This is basically the same as declaring arrays in C (say).

You could workaround by above some size using the dynamic size interface that Eigen provides (then they'll be heap allocated).

**Comment by miklos1** [comment]
Related: https://github.com/firedrakeproject/tsfc/issues/127

**Comment by pefarrell** [comment]
I've just bumped into this again. Is there any prospect of a fix?

**Comment by pefarrell** [comment]
Temporary workaround: `export PYOP2_CFLAGS="-DEIGEN_STACK_ALLOCATION_LIMIT=0"`

**Comment by wence-** [comment]
This is fixed by migration to loopy. Plausibly for very big local tensors one still wants to heap allocate...
